parsers/tidb/tidb.go

```diff
@@ -0,0 +1,448 @@
+	"reflect"
+	"regexp"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/honeycombio/honeytail/event"
+	"github.com/honeycombio/honeytail/parsers"
+	"github.com/pingcap/tidb/sessionctx/variable"
+	"github.com/pingcap/tidb/util/execdetails"
+	"github.com/pingcap/tidb/util/hack"
+	"github.com/pingcap/tidb/util/plancodec"
+	"github.com/sirupsen/logrus"
+)
+
+const timeFormat = "2006-01-02 15:04:05.999999999"
+
+type Options struct {
+	NumParsers int `hidden:"true" description:"number of TIDB parsers to spin up" yaml:"-"`
+}
+
+// Parser is used to read slow log data.
+type Parser struct {
+	outputCols            []string
+	initialized           bool
+	columnValueFactoryMap map[string]slowQueryColumnValueFactory
+	SampleRate            int
+	InstanceAddr          string
+	InstanceHostName      string
+	conf                  Options
+	wg                    sync.WaitGroup
+}
+
+// kvSplitRegex: it was just for split "field: value field: value..."
+var kvSplitRegex = regexp.MustCompile(`\w+: `)
+
+// splitByColon split a line like "field: value field: value..."
+func splitByColon(line string) (fields []string, values []string) {
+	matches := kvSplitRegex.FindAllStringIndex(line, -1)
+	fields = make([]string, 0, len(matches))
+	values = make([]string, 0, len(matches))
+
+	beg := 0
+	end := 0
+	for _, match := range matches {
+		// trim ": "
+		fields = append(fields, line[match[0]:match[1]-2])
+
+		end = match[0]
+		if beg != 0 {
+			// trim " "
+			values = append(values, line[beg:end-1])
+		}
+		beg = match[1]
+	}
+
+	if end != len(line) {
+		// " " does not exist in the end
+		values = append(values, line[beg:])
+	}
+	return fields, values
+}
+
+func GetInterfaceIpv4Addr(interfaceName string) (addr string, err error) {
+	var (
+		ief      *net.Interface
+		addrs    []net.Addr
+		ipv4Addr net.IP
+	)
+	if ief, err = net.InterfaceByName(interfaceName); err != nil { // get interface
+		return
+	}
+	if addrs, err = ief.Addrs(); err != nil { // get addresses
+		return
+	}
+	for _, addr := range addrs { // get ipv4 address
+		if ipv4Addr = addr.(*net.IPNet).IP.To4(); ipv4Addr != nil {
+			break
+		}
+	}
+	if ipv4Addr == nil {
+		return "", errors.New(fmt.Sprintf("interface %s don't have an ipv4 address\n", interfaceName))
+	}
+	return ipv4Addr.String(), nil
+}
+
+func (p *Parser) Init(options interface{}) error {
+	//sctx := mock.NewContext()
+	if !p.initialized {
+		if err := p.initialize(); err != nil {
+			return err
+		}
+	}
+	typ := reflect.TypeOf(options) // 类型
+
+	if typ.Kind() == reflect.Ptr {
+		p.conf = *options.(*Options)
+	} else {
+		p.conf = options.(Options)
+	}
+
+	ip, err := GetInterfaceIpv4Addr("eth0")
+	if err != nil {
+		logrus.WithError(err).Warn("failed to get eth0 ipv4 address")
+		return err
+	}
+	p.InstanceAddr = ip
+
+	name, err := os.Hostname()
+	if err != nil {
+		logrus.WithError(err).Warn("failed to get hostname")
+		return err
+	}
+	p.InstanceHostName = name
+
+	return nil
+}
+
+func (p *Parser) initialize() error {
```

Comment:
This could be part of the main `Init` func. Separting it and using an `initialized` field to check if it's already done isn't neccesary. Init should only be called once anyway.